### PR TITLE
Add flag link to profile dropdown view

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [x] Feature

## Description

This PR adds a "flag user" option — for usage by trusted members — directly to the user profile.

## Related Tickets & Documents

https://github.com/forem/rfcs/pull/73

### Screenshots

**Flagging:**

image
image
image

**Unflagging:**

image
image
image

## QA Instructions

**Flagging scenario 1:**

1. Go to an unflagged user's profile
2. Click "Flag @username" in the dropdown
3. Click "Cancel" in the alert window

Noting should happen.

**Flagging scenario  2:**
1. Go to an unflagged user's profile
2. Click "Flag @username" in the dropdown
3. Click "Ok" in the alert window

A new alert with the message "All posts by this author will be less visible." should pop up.  A reaction of type "vomit" should have been recorded in the DB.

**Unlagging scenario 1:**

1. Go to a flagged user's profile
2. Click "Unflag @username" in the dropdown
3. Click "Cancel" in the alert window

Noting should happen.

**Unflagging scenario  2:**
1. Go to a flagged user's profile
2. Click "Unflag @username" in the dropdown
3. Click "Ok" in the alert window

A new alert with the message "Unflagged the author, the visibility of their posts has been restored." should pop up.  The "vomit" reaction should be deleted from the DB.

Please test flagging and then unflagging both while staying on the page and also with navigating away in between.

### UI accessibility concerns?

n/a

## Added tests?

- [x] Yes, but very basic. I'm not good at frontend testing. Should we add system specs for this at least?

## [Forem core team only] How will this change be communicated?

- [X] I will share this change internally with the appropriate teams

### CODE REVIEW

**Comment by citizen428** [comment]
One thing I noticed while working on this: repeatedly flagging the user actually will toggle the flag, due to this code in `ReactionsController`:

```ruby
  def create
    ...
    reaction = Reaction.where(
      user_id: current_user.id,
      reactable_id: params[:reactable_id],
      reactable_type: params[:reactable_type],
      category: category,
    ).first

    # if the reaction already exists, destroy it
    if reaction
      result = handle_existing_reaction(reaction)
    else
      ....
    end
  end

  def handle_existing_reaction(reaction)
    result = destroy_reaction(reaction)

    if reaction.negative? && current_user.auditable?
      updated_params = params.dup
      updated_params[:action] = "destroy"
      Audit::Logger.log(:moderator, current_user, updated_params)
    end

    result
  end
```

This is also true for the current mod center version and there's no indication that this is the case. So if a user forgets that they already flagged someone, they may inadvertently undo it while believing that they flagged the user.

**Comment by rhymes** [comment]
@citizen428 

Good catch! I think that from a UX perspective we should either store the connection “user A flagged user B” so that we can disable the button for them after the first successful flagging or make it so that multiple “flaggings” don’t act as a on-off switch.

The peril with the first solution is that a person might be flagged multiple times over the lifetime of their membership. If I flag a user today and then they need to be flagged six month from now again, I wouldn’t want the button to be disabled. Counter argument: people might “spam flag” repeatedly members they don’t like, though I don’t see this happening as flagging isn’t an anonymous operation.

For these reasons I’d be inclined to vote for changing the feature such as a two flags have the same effect as one but the @forem/product team might have further insights.

**Comment by citizen428** [comment]
@vaidehijoshi What do you think about the options outlined above? I think this needs a product decision as it won't only change the behavior of this button, but also the existing one in the mod center. So unless we know how exactly we want to handle this case, I'll move this issue to blocked on our cycle board.

**Comment by vaidehijoshi** [comment]
Sorry that I was slow to reply here! 😓 

I definitely agree with both of your assessments that repeatedly clicking the button shouldn't toggle the "flagged" status.

I thought a lot about this, specifically about whether we want to allow a mod user to flag the same person multiple times or not. On the one hand, we may want to allow that so that we can empower mods to repeatedly flag problematic behavior. On the other hand, if a mod flags something, one would hope that the Forem admin(s) would see the mod's flag and take some action on it (banish, suspension, etc.).

And honestly...either of the two above ^^ options would be better than what we currently have (which apparently is toggle-able flagging?) 🙃 

Here's what I'm thinking we should do: let's definitely change these buttons so that they aren't "toggling" the flagged status. Instead, I think we should allow a mod to create multiple flaggings (`make it so that multiple “flaggings” don’t act as a on-off switch.`, as @rhymes put it).

Ideally, I think it might also be nice to indicate to the mod that they _already_ flagged someone and ask whether they are sure they want to do it again (something like `You flagged this user on March 1, 2021. Are you sure you want to flag them again?`). But that might be more information than is actually necessary, and doesn't necessarily need to be done in the context of this PR. I think we should trust our mods in general, but building in UX that prompts them to think about multiple flaggings might help promote more compassionate behavior :)

File: app/javascript/profileDropdown/flagButton.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/* eslint-disable no-alert */
+export function initFlag() {
```
Comment (citizen428) [inline app/javascript/profileDropdown/flagButton.js:2]: This whole file was modeled closely after `blockButton.js` so we handle the two cases consistently.

File: app/controllers/reactions_controller.rb [hunk 2]
```diff
@@ -73,7 +73,9 @@ def create
     ).first
 
     # if the reaction already exists, destroy it
-    if reaction
+    # NOTE: @citizen Temporarily we ingore vomit reactions, to avoid accidental
```
Comment (citizen428) [inline app/controllers/reactions_controller.rb:76]: @rhymes @vaidehijoshi This is by no means an elegant solution, but I think I want to keep the bigger refactoring out of the scope of this PR/RFC. What do you think?

File: spec/requests/reactions_spec.rb [hunk 3]
```diff
@@ -190,17 +190,33 @@
     context "when reacting to an article" do
       before do
         sign_in user
-        post "/reactions", params: article_params
```
Comment (citizen428) [inline spec/requests/reactions_spec.rb:193]: I removed this and updated the existing specs because `change` assertions are generally a bit more robust than asserting on specific counts. I also dislike specs that are only assertions without any actual actions in them because they are hidden away in `before` blocks.

File: spec/requests/reactions_spec.rb [hunk 4]
```diff
@@ -190,17 +190,33 @@
...
-        # same route to destroy, so sending POST request again
         post "/reactions", params: article_params
-        expect(Reaction.all.size).to eq(0)
+        expect do
+          # same route to destroy, so sending POST request again
+          post "/reactions", params: article_params
+        end.to change(Reaction, :count).by(-1)
+      end
+
+      it "does not destroy existing vomit reactions" do
```
Comment (citizen428) [inline spec/requests/reactions_spec.rb:209]: This is new and covers the quickfix.

File: spec/requests/reactions_spec.rb [hunk 5]
```diff
@@ -270,17 +286,20 @@
...
       it "creates reaction" do
-        expect(Reaction.last.reactable_id).to eq(user.id)
+        expect do
+          post "/reactions", params: user_params
+        end.to change(Reaction, :count).by(1)
       end
 
-      it "destroys existing reaction" do
-        # same route to destroy, so sending POST request again
+      it "does not destroy an existing reaction" do
```
Comment (citizen428) [inline spec/requests/reactions_spec.rb:297]: This is was changed to cover the quickfix.

File: spec/requests/reactions_spec.rb [hunk 6]
```diff
@@ -270,17 +286,20 @@
     context "when vomiting on a user" do
       before do
         sign_in trusted_user
-        post "/reactions", params: user_params
```
Comment (citizen428) [inline spec/requests/reactions_spec.rb:273]: Changed for the same reasons as above.

**Comment by citizen428** [comment]
@vaidehijoshi @rhymes I updated the code with a quick fix now. Do you think it's acceptable to ignore repeated flagging for the purpose of this PR/RFC and then revisit the proper refactoring in a separate PR?

Also, what do you think in terms of frontend testing? The somewhat related block button doesn't seem to have any dedicated tests, or I'm just unable to find them.

**Comment by lisasy** [comment]
@vaidehijoshi @rhymes @citizen428 

**I agree with Vaidehi that it is sensible to allow multiple flagging**. Note that the entirety of the UX comprises of a few things:

1. Mod intends to flag a user
2. Mod opens dropdown and sees a button saying "Flag {user}"
3. Mod clicks on button and trigger modal to confirm flagging
4. Mod receives feedback that this successfully happened

If this PR is scoped narrowly to include the "Flag {user}" link in the dropdown, we can have that link but not worry about toggling changing the text to "Already flagged". I think that future iterations of this flow where we clarify to the mod that they have *already* flagged someone (and we haven't address this yet) can be messaged within the modal.

**Comment by citizen428** [comment]
@lisasy This is what the current version is doing, it silently ignores subsequent flags. I figured getting this one out fast and then properly thinking about a follow-up would be better than rushing something into this PR.

**Comment by citizen428** [comment]
While getting the specs to pass I noticed one place where we may actually want the vomit reactions to be toggle-able, which is in the notifications:

image

I update the spec to not check the toggling of the vomit reaction but it's something we need to keep in mind when working on a proper solution.

To summarize: 
1. Toggle reactions are currently (before this PR) toggleable. 
2. The only place where we communicate this back to the user seems to be via the icons in the notifications mod UI, not in other places like the mod center or this newly added link.
3. In the future, we need to think about this more holistically:  do we want to unflag? Allow multiple flaggings with aggregate penalties, etc.?

I'll put up a stage 0 RFC for this, even I may not be the person working on it.

File: app/controllers/stories_controller.rb [hunk 7]
```diff
@@ -236,6 +236,7 @@ def handle_user_index
     # 2nd with 1 badge (!) <-- and that would look off.
     @badges_limit = 6
     @profile = @user.profile.decorate
+    @flag_status = Reaction.where(user_id: session_current_user_id, reactable: @user).any?
```
Comment (citizen428) [inline app/controllers/stories_controller.rb:239]: Just to double-check, is this safe to use with our caching?

**Comment by lisasy** [comment]
@citizen428 that sounds like a sound approach to me... and also, I think us defining this behavior and making it has not only consistent, but *good* UX coverage across the board is something we'll need.

File: app/javascript/profileDropdown/flagButton.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  if (user.id === parseInt(profileUserId, 10) || !user.trusted) {
+    flagButton.remove();
+  }
+
+  function flag() {
+    const confirmFlag = window.confirm(
+      flagStatus
+        ? "Are you sure you want to unflag this person? This will restore their posts' visibility."
+        : 'Are you sure you want to flag this person? This will make all of their posts less visible.',
+    );
```
Comment (vaidehijoshi) [inline app/javascript/profileDropdown/flagButton.js:28]: I thought about this while QAing this and then went back to the original RFC to look at the [definition of done](https://github.com/forem/rfcs/pull/73/files#diff-5dc7721fc901e825ee9a52b83cb6c4cb07c72c6df4286f97fbdb76fdba4f6408R31-R33)...

It feels a bit weird to me to flag a user and have to click through _two_ confirmation buttons. I know that @lisasy mentioned in [this comment](https://github.com/forem/forem/pull/12862#issuecomment-796875917) that we the mod should receive _"feedback that this [the flagging] successfully happened"_, but I think we might want to make this a more _passive_ notice rather than an "active" one where the mod has to click twice to flag someone. The first click (the actual "confirmation") makes sense to be more active, as we definitely want to ensure that the mod knows the implication of what they're about to do.

But after that, there isn't any _action_ to be taken, so the browser alert feels kind of heavyweight for this:
Screen Shot 2021-03-26 at 11 34 56 AM

This is just the feeling I got when QAing the flagging feature; I realize that the RFC doesn't specify what _kind_ of confirmation should be used here 😓 Could we maybe just show the second confirmation as a flash message of sorts (something that doesn't require another click from the mod?)?

@lisasy What do you think?

File: app/javascript/profileDropdown/flagButton.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  let flagStatus = flagButton.dataset.flagStatus === 'true';
+
+  if (user.id === parseInt(profileUserId, 10) || !user.trusted) {
+    flagButton.remove();
+  }
+
+  function flag() {
+    const confirmFlag = window.confirm(
+      flagStatus
+        ? "Are you sure you want to unflag this person? This will restore their posts' visibility."
```
Comment (vaidehijoshi) [inline app/javascript/profileDropdown/flagButton.js:26]:
```suggestion
        ? "Are you sure you want to unflag this person? This will make all of their posts visible again."
```

File: app/javascript/profileDropdown/flagButton.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
+/* eslint-disable no-alert */
+export function initFlag() {
```
Comment (vaidehijoshi) [inline app/javascript/profileDropdown/flagButton.js:2]: Could we add some brief documentation for these functions? [`initializeBroadcast.js`](https://github.com/forem/forem/blob/master/app/assets/javascripts/initializers/initializeBroadcast.js) has some good JSDoc examples we could follow! 😃

File: app/controllers/stories_controller.rb (hunk 7 above)
Comment (vaidehijoshi) [reply app/controllers/stories_controller.rb:239]: I _think_ so? I poked around and it seems like `session_current_user_id` is used in a few different controllers, and we have [this check](https://github.com/forem/forem/blob/master/app/controllers/concerns/edge_cache_safety_check.rb) in place 🤔

**Review by vaidehijoshi — COMMENTED** [review]
> Should we add system specs for this at least?

+100! I think a system spec would be great here, just to test that the flag button does in fact render and is clickable when we expect and can create the vomit reaction when we expect 👍🏽

File: app/javascript/profileDropdown/flagButton.js (hunk 8 above)
Comment (citizen428) [reply app/javascript/profileDropdown/flagButton.js]: Right, I'm not a big fan of the second confirmation either. I'm not sure how we usually do flash messages in async requests from the frontend, but I can have a look.

File: app/controllers/stories_controller.rb (hunk 7 above)
Comment (citizen428) [reply app/controllers/stories_controller.rb:239]: Yes, that's the conclusion I had reached too, but when it comes to our caching I usually like to double or triple check.

File: app/javascript/profileDropdown/flagButton.js (hunk 10 above)
Comment (citizen428) [reply app/javascript/profileDropdown/flagButton.js:2]: Yeah, sure.

File: app/javascript/profileDropdown/flagButton.js (hunk 8 above)
Comment (citizen428) [reply app/javascript/profileDropdown/flagButton.js]: I just removed the second popup now. The menu item in the existing popup will immediately change from `Flag` to `Unflag` and vice versa, and while potentially a bit too subtle I think it may be enough. Definitely better than the extra popup.

**Comment by citizen428** [comment]
@vaidehijoshi Thanks for the feedback, I

1. accepted your suggestion for the copy change,
2. added some JSDoc,
3. removed the second confirmation pop-up,
4. restructured the view a bit, both block and flag buttons will now only show for logged-in users. Before they always showed but didn't have any behavior attached, and
5. added a system spec that took way too long to write because async and Capybara... 😢 

In the end, I settled for a slightly simpler system spec than I wanted because of the above-mentioned async issues. But I think it may be good enough. Generally speaking, if the link is there, it will work unlike before, where the link may have been there but without any actual behavior.

**Comment by citizen428** [comment]
I now have two failing JS specs about mention autocomplete and I'm not sure how this happened... Should I merge `master` back here or does anyone else know what's going on?

File: app/views/users/show.html.erb [hunk 11]
```diff
@@ -76,8 +85,8 @@
             <span class="profile-header__meta__item -ml-1">
               <% Profile.special_social_link_attributes.each do |attribute| %>
                 <% if @user.send(attribute).present? %>
```
Comment (rhymes) [inline app/views/users/show.html.erb:87]: interesting that Rubocop didn't catch this one, we're using `.send` here and `.public_send` on the following line

**Review by rhymes — APPROVED** [review]
Tested manually 🆗

**Review by vaidehijoshi — APPROVED** [review]
Thanks for the new specs and confirmation flow changes, @citizen428! :)

File: app/views/users/show.html.erb (hunk 11 above)
Comment (citizen428) [reply app/views/users/show.html.erb:87]: This isn't really related to this PR, but I can change it if you want?

File: app/views/users/show.html.erb [hunk 12]
```diff
@@ -75,9 +84,9 @@
             <% end %>
             <span class="profile-header__meta__item -ml-1">
               <% Profile.special_social_link_attributes.each do |attribute| %>
-                <% if @user.send(attribute).present? %>
-                  <a href="<%= @user.send(attribute) %>" target="_blank" rel="noopener me" class="px-1 align-middle inline-block">
-                    <%= inline_svg_tag("#{attribute.gsub("_url","")}.svg", class: "crayons-icon", aria: true, title: "#{attribute.split("_").first.titleize} logo") %>
+                <% if @user.public_send(attribute).present? %>
```
Comment (citizen428) [inline app/views/users/show.html.erb:87]: Unrelated to this PR, noticed by @rhymes-ocop.

**Comment by citizen428** [comment]
Will leave open a bit longer to give @aitchiss a chance to look at this.

File: app/javascript/profileDropdown/flagButton.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  const flagButton = document.getElementById(
+    'user-profile-dropdownmenu-flag-button',
+  );
+
+  if (!flagButton) {
+    // button not always present when this is called
+    return;
+  }
+
+  /* eslint-disable-next-line no-undef */
```
Comment (aitchiss) [inline app/javascript/profileDropdown/flagButton.js:17]: suggestion (non-blocking) - if you add

```
/* global userData */
```

to the top of the file, you can remove the `disable-next-line` here

File: app/javascript/profileDropdown/flagButton.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+  const user = userData();
+  if (!user) {
+    return;
+  }
+
+  const { profileUserId, profileUserName } = flagButton.dataset;
+  let flagStatus = flagButton.dataset.flagStatus === 'true';
```
Comment (aitchiss) [inline app/javascript/profileDropdown/flagButton.js:24]: (not-blocking) - might just be me, but I think naming this something like `isUserFlagged` would make things clearer

**Review by aitchiss — COMMENTED** [review]
It works well! I think there some improvements we could make to that overflow menu generally, but it's not really part of this piece of work (future fun 🙂).

It would be great to have a Cypress test for this rather than the rspec one, as it could test the full E2E process - I could help with that if you like?

File: app/javascript/profileDropdown/flagButton.js (hunk 13 above)
Comment (citizen428) [reply app/javascript/profileDropdown/flagButton.js:17]: Ah, thanks for this, I had just copied this over from `blockButton.js`, but prefer your solution.

File: app/javascript/profileDropdown/flagButton.js (hunk 14 above)
Comment (citizen428) [reply app/javascript/profileDropdown/flagButton.js:24]: I have no strong preference so you saying you find it clearer is enough of an argument to rename it for me. Thanks for the suggestion 👍

**Comment by citizen428** [comment]
@aitchiss Thanks for the feedback, I addressed both of your suggestions. 

Regarding the e2e test: I haven't yet had much time to properly acquaint myself with Cypress 😭 Due to this PR sitting around for a longish time I started some other work which includes this major refactoring PR: https://github.com/forem/forem/pull/13095

I'm unsure if I really want to interrupt this right now to teach myself Cypress but if you're willing to help that'd be great!

**Review by msarit — APPROVED** [review]
LGTM! 🚀  Great work everyone!
@aitchiss Let me know if I can help out with the E2E tests 😄

**Comment by aitchiss** [comment]
> @aitchiss Thanks for the feedback, I addressed both of your suggestions.
> 
> Regarding the e2e test: I haven't yet had much time to properly acquaint myself with Cypress 😭 Due to this PR sitting around for a longish time I started some other work which includes this major refactoring PR: #13095
> 
> I'm unsure if I really want to interrupt this right now to teach myself Cypress but if you're willing to help that'd be great!

Sure! I had a quick look at it earlier and it's not quite as straightforward as I thought (there's some strange caching happening between test runs). It'll be good to look into more, as they're things that will need solved for future Cypress tests anyway, but I'm on cooldown this week and realistically won't get to it until later next week.

TL;DR - given this has been open for a while, I'm happy if you want to merge without the Cypress tests for now and I can always follow up with them later 👍

**Comment by citizen428** [comment]
Thanks @aitchiss, that sounds like a good solution.
